File: xr0-rs/src/state/static.rs
use std::ptr;

use crate::state::block::{block_create, block_str};
use crate::state::location::location_copy;
use crate::util::{dynamic_str, strbuilder_build, strbuilder_create, Map, OwningCStr};
use crate::{cstr, strbuilder_write, Block, Location, StrBuilder};

pub struct StaticMemory {
    pub blocks: Vec<Box<Block>>,
    pub pool: Box<Map>,
}

pub unsafe fn static_memory_create() -> *mut StaticMemory {
    Box::into_raw(Box::new(StaticMemory {
        blocks: vec![],
        pool: Map::new(),
    }))
}

pub unsafe fn static_memory_destroy(sm: *mut StaticMemory) {
    // Note: Original leaked the StaticMemory itself.
    drop(Box::from_raw(sm));
}

pub unsafe fn static_memory_str(sm: *mut StaticMemory, indent: *mut libc::c_char) -> OwningCStr {
    let b: *mut StrBuilder = strbuilder_create();
    for (i, block) in (*sm).blocks.iter().enumerate() {
        strbuilder_write!(b, "{}{i}: {}\n", cstr!(indent), block_str(block));
    }
    strbuilder_build(b)
}

pub unsafe fn static_memory_copy(sm: *mut StaticMemory) -> *mut StaticMemory {
    Box::into_raw(Box::new((*sm).clone()))
}

impl Clone for StaticMemory {
    fn clone(&self) -> Self {
        unsafe {
            StaticMemory {
                blocks: self.blocks.clone(),
                pool: pool_copy(&self.pool),
            }
        }
    }
}

unsafe fn pool_copy(p: &Map) -> Box<Map> {
    let mut pcopy = Map::new();
    for (k, v) in p.pairs() {
        pcopy.set(
            dynamic_str(k),
            location_copy(&*(v as *mut Location)) as *const libc::c_void,
        );
    }
    pcopy
}

pub unsafe fn static_memory_newblock(sm: *mut StaticMemory) -> libc::c_int {
    let address = (*sm).blocks.len() as libc::c_int;
    (*sm).blocks.push(block_create());
    address
}

pub unsafe fn static_memory_getblock(sm: *mut StaticMemory, address: libc::c_int) -> *mut Block {
    if address as usize >= (*sm).blocks.len() {
        return ptr::null_mut();
    }
    &mut *(*sm).blocks[address as usize]
}

pub unsafe fn static_memory_stringpool(
    sm: *mut StaticMemory,
    lit: *mut libc::c_char,
    loc: *mut Location,
) {
    (*sm).pool.set(
        dynamic_str(lit),
        location_copy(&*loc) as *const libc::c_void,
    );
}

pub unsafe fn static_memory_checkpool(
    sm: *mut StaticMemory,
    lit: *mut libc::c_char,
) -> *mut Location {
    (*sm).pool.get(lit) as *mut Location
}

File: src/state/static.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <assert.h>

#include "ast.h"
#include "location.h"
#include "block.h"
#include "value.h"
#include "util.h"

struct static_memory {
	struct block_arr *blocks;	
	struct map *pool;
};

struct static_memory *
static_memory_create(void)
{
	struct static_memory *sm = malloc(sizeof(struct static_memory));
	assert(sm);
	sm->blocks = block_arr_create();
	sm->pool = map_create(); 
	return sm;
}

void
static_memory_destroy(struct static_memory *sm)
{
	block_arr_destroy(sm->blocks);
}

char *
static_memory_str(struct static_memory *sm, char *indent)
{
	struct strbuilder *b = strbuilder_create();
	int n = block_arr_nblocks(sm->blocks);
	struct block **arr = block_arr_blocks(sm->blocks);
	for (int i = 0; i < n; i++) {
		char *block = block_str(arr[i]);
		strbuilder_printf(b, "%s%d: %s\n", indent, i, block);
		free(block);
	}
	return strbuilder_build(b);
}

static struct map *
pool_copy(struct map *);

struct static_memory *
static_memory_copy(struct static_memory *sm)
{
	struct static_memory *copy = malloc(sizeof(struct static_memory));	
	copy->blocks = block_arr_copy(sm->blocks);
	copy->pool = pool_copy(sm->pool);
	return copy;
}

static struct map *
pool_copy(struct map *p)
{
	struct map *pcopy = map_create();
	for (int i = 0; i < p->n; i++) {
		struct entry e = p->entry[i];
		map_set(
			pcopy,
			dynamic_str(e.key),
			location_copy((struct location *) e.value)
		);
	}
	return pcopy;
}

int
static_memory_newblock(struct static_memory *sm)
{
	int address = block_arr_append(sm->blocks, block_create(1));

	int n = block_arr_nblocks(sm->blocks);
	assert(n > 0);
	return address;
}

struct block *
static_memory_getblock(struct static_memory *sm, int address)
{
	if (address >= block_arr_nblocks(sm->blocks)) {
		return NULL;
	}
	return block_arr_blocks(sm->blocks)[address];
}

/* string pooling to write one string literal to static_memory and all subsequent
 * literals reference it is used to avoid infinite loops in splitting logic with
 * different literals per selection condition */
void
static_memory_stringpool(struct static_memory *sm, char *lit, struct location *loc)
{
	map_set(
		sm->pool,
		dynamic_str(lit),
		location_copy(loc)
	);
}

struct location *
static_memory_checkpool(struct static_memory *sm, char *lit)
{
	return (struct location *) map_get(sm->pool, lit);
}
